Ha1 password option

ps opened the pull request:
Added the ability to pass a flag to HttpDigestAuth in order to use 'ha1' combination of username + realm + password as the password stored for a user. I added this feature so that when storing users in the database, the application does not have to store the actual plain text password.

miguelgrinberg commented:
What is the purpose of storing the generated hash versus calculating it? Do you want to improve performance, or do you want to avoid storing plain text passwords in your database?

If this is about performance, you aren't really saving much. The calculation requires 3 MD5s, and with this change you are removing one of them. Plus this introduces the need to ensure this MD5 hash remains in sync when the username or the password is changed.

If you want to avoid storing plain text passwords in your database, then I think a better solution is to hash the password and store that instead. Then on the client side, hash the password in the same way before entering into the digest algorithm.

ps commented:
The purpose was to avoid storing a plain text password. And yes I could hash on the client side, but I added this feature so that client side would not have to worry about hashing. 

This was just an adjustment that I made to fit for a small project of mine. I figured I'd submit a PR to see what you think about it. I can modify this a bit more if you have some suggestions, but in the end the decision is yours on this.

miguelgrinberg commented:
My main concern is that this breaks encapsulation. The application now needs to know how to encode this hash, and needs to keep it updated when the username or the password change.

Could I ask you to add the ha1 encoding to the extension, so that an application can use it to generate the hash that goes in the database? Something like what you did for the tests, but exposed by the extension as a service to applications. I think that will make it a bit more reasonable.

ps commented:
I added a 'generate_ha1' function to HTTPDigestAuth class. Is this what you meant or should I have added it somewhere else?

Again, I won't be offended if you are not comfortable with this addition to the existing code.
